soroban-env-host/src/storage.rs

```diff
@@ -154,6 +154,34 @@ pub struct Storage {
 // Notes on metering: all storage operations: `put`, `get`, `del`, `has` are
 // covered by the underlying [MeteredOrdMap] and the [Footprint]'s own map.
 impl Storage {
+    /// Only a subset of Stellar's XDR ledger key or entry types are supported
+    /// by Soroban: accounts, trustlines, contract code and data. The rest are
+    /// never used by stellar-core when interacting with the Soroban host, nor
+    /// does the Soroban host ever generate any. Therefore the storage system
+    /// will reject them with [ScErrorCode::InternalError] if they ever occur.
+    pub fn check_supported_ledger_entry_type(le: &LedgerEntry) -> Result<(), HostError> {
+        use crate::xdr::LedgerEntryData::*;
+        match le.data {
+            Account(_) | Trustline(_) | ContractData(_) | ContractCode(_) => Ok(()),
+            Offer(_) | Data(_) | ClaimableBalance(_) | LiquidityPool(_) | ConfigSetting(_)
+            | Ttl(_) => Err((ScErrorType::Storage, ScErrorCode::InvalidInput).into()),
```

Comment:
We don't expect invalid entries to get into host, do we? So maybe internal error would be more appropriate. I don't have a strong opinion though.